atlasdb-timelock-server/src/test/java/com/palantir/atlasdb/timelock/paxos/PaxosTimestampBoundStoreTest.java

```diff
@@ -0,0 +1,280 @@
+                    PaxosAcceptor.class,
+                    acceptor,
+                    failureController,
+                    EXCEPTION));
+            PaxosLearner learner = PaxosLearnerImpl.newLearner(LEARNER_DIR_PREFIX + i);
+            learners.add(ToggleableExceptionProxy.newProxyInstance(
+                    PaxosLearner.class,
+                    learner,
+                    failureController,
+                    EXCEPTION));
+            failureToggles.add(failureController);
+        }
+
+        store = createPaxosTimestampBoundStore(0);
+    }
+
+    @After
+    public void tearDown() throws InterruptedException, IOException {
+        try {
+            executor.shutdownNow();
+            boolean terminated = executor.awaitTermination(10, TimeUnit.SECONDS);
+            if (!terminated) {
+                throw new IllegalStateException(STILL_RUNNING_MESSAGE);
+            }
+        } finally {
+            FileUtils.deleteDirectory(new File(LOG_DIR));
+        }
+    }
+
+    @Test
+    public void timestampsBeginFromZero() {
+        assertThat(store.getUpperLimit()).isEqualTo(0L);
+    }
+
+    @Test
+    public void canStoreUpperLimit() {
+        store.storeUpperLimit(TIMESTAMP_1);
+        assertThat(store.getUpperLimit()).isGreaterThanOrEqualTo(TIMESTAMP_1);
+    }
+
+    @Test
+    public void throwsIfStoringLimitLesserThanUpperLimit() {
+        store.storeUpperLimit(TIMESTAMP_2);
+        assertThat(store.getUpperLimit()).isGreaterThanOrEqualTo(TIMESTAMP_2);
+        assertThatThrownBy(() -> store.storeUpperLimit(TIMESTAMP_1)).isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    public void canOperateWithMinorityOfNodesDown() {
+        failureToggles.get(1).set(true);
+        failureToggles.get(2).set(true);
+        store.storeUpperLimit(TIMESTAMP_1);
+        assertThat(store.getUpperLimit()).isGreaterThanOrEqualTo(TIMESTAMP_1);
+    }
+
+    @Test
+    public void throwsIfCannotObtainQuorum() {
+        failureToggles.get(1).set(true);
+        failureToggles.get(2).set(true);
+        failureToggles.get(3).set(true);
+        assertThatThrownBy(() -> store.getUpperLimit()).isInstanceOf(ServiceNotAvailableException.class);
+    }
+
+    @Test
+    public void retriesProposeUntilSuccessful() throws Exception {
+        store = createOnceFailingPaxosTimestampBoundStore(0);
+        store.storeUpperLimit(TIMESTAMP_1);
+        assertThat(store.getUpperLimit()).isGreaterThanOrEqualTo(TIMESTAMP_1);
+    }
+
+    @Test
+    public void throwsIfBoundUnexpectedlyChangedUnderUs() {
+        PaxosTimestampBoundStore additionalStore = createPaxosTimestampBoundStore(1);
+        additionalStore.storeUpperLimit(TIMESTAMP_1);
+        assertThatThrownBy(() -> store.storeUpperLimit(TIMESTAMP_2))
+                .isInstanceOf(MultipleRunningTimestampServiceError.class);
+    }
+
+    @Test
+    public void canReadStateFromDistributedLogs() {
```

Comment:
Yep, that's right - the triggering condition for MRTSE is actually when you try to update the bound and your own bound is not synced up.